ENH: stats: Add matrix variate t distribution

johnmdusel opened the pull request:
#### Reference issue
Closes  [issue #22870](https://github.com/scipy/scipy/issues/22870).

#### What does this implement/fix?
Background explanation in [this post on the developer forum](https://discuss.scientific-python.org/t/feature-addition-adding-matrix-t-distribution-to-scipy-stats/1845)

#### Additional information

From the PR checklist:

- This code can be distributed under a BSD license. I followed the existing random matrix implementation for data validation and wrote the methods from scratch.

- I provided unit tests in `stats.tests.test_multivariate.TestMatrixT`. These are mostly simiar to the ones for `matrix_normal`, except that the moment checking and sample checking also perform a spot-check of two basic theorems from [Gupta and Nagar (2000)](https://www.taylorfrancis.com/books/mono/10.1201/9780203749289/matrix-variate-distributions-nagar-gupta). All unit tests pass locally.

- Re: style: I used `black` to format my additions to `stats._multivariate`. For unit tests and benchmarking I followed the style of existing code.

- I provided benchmarks in `stats.MatrixSampling`. This also includes benchmarks for `rvs` methods of `matrix_normal` and `invwishart`.

- All public functions have docstrings with examples, and I've verified that the documentation builds locally and renders correctly.

dschmitz89 committed:
TST: add tests for ncfdtri

Commit:
ENH: stats: Add matrix variate t-distribution code to `_multivariate`

Commit:
BUG: Correct handling of missing degrees-of-freedom parameter. Ensure that frozen instance checks for singularity of spread matrices.

Commit:
TST: stats.tests.test_multivariate: Add `TestMatrixT` class
    - Completed `test_bad_input`
    - Stubs of other tests

Commit:
TST/BUG: stats: (WIP) Update unit tests and fix bugs.

    - Adds unit tests for bad input and default inputs and covariance expansion to
      matrix variate t test class.
    - Fixes bugs in matrix variate t distribution which were exposed by those tests.

Commit:
TST/BUG: stats: (WIP) Update unit tests and fix bugs.

    - Adds unit tests for array input to matrix variate t test class.

    - Fixes bugs in matrix variate t distribution which were exposed by those tests.

        _logpdf did not expect to receive an array input with > 3 dimensions and its Einstein sum
        is defined based on that expectation. So, if there are > 3 dimensions then _logpdf will
        flatten out the extra ones, perform the calculation, then reshape before returning.

Commit:
Finish adding analogues of unit tests from matrix normal.

Commit:
BENCH: Add matrix variate sampling benchmark.

Covers matrix normal, matrix t, and inverse Wishart.

Commit:
STY: Remove extra whitespace.

Commit:
DOC: (WIP) Resolving Sphinx build error

Commit:
DOC: stats: Fixes from local docs build.

Commit:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

Commit:
MAINT: stats: Fix docstring typo.

tylerjereddy commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,631 @@ def entropy(self):
+            if len(meanshape) != 2:
+                raise ValueError("Array `mean` must be 2D.")
+            if np.any(meanshape == 0):
+                raise ValueError("Array `mean` has invalid shape.")
```
Lacks test coverage if it matters.

johnmdusel replied:
Added coverage in b1916d6.

tylerjereddy commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,631 @@ def entropy(self):
+        if rowshape[0] != rowshape[1]:
+            raise ValueError("Array `rowcov` must be square.")
+        if rowshape[0] == 0:
+            raise ValueError("Array `rowcov` has invalid shape.")
```
The 3 error code paths above are not covered by tests, in case it matters.

johnmdusel replied:
Added coverage in b1916d6.

tylerjereddy commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,631 @@ def entropy(self):
+            raise ValueError("Array `rowcov` has invalid shape.")
+        numrows = rowshape[0]
+        if np.isnan(rowcov).any():
+            import pdb; pdb.set_trace()
```
I suspect we don't want to drop the user into the debugger here--residual debugging code to clean up?

johnmdusel replied:
Resolved in e45cb56. Thanks for catching this oversight.

tylerjereddy commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,631 @@ def entropy(self):
+        if colshape[0] != colshape[1]:
+            raise ValueError("Array `colcov` must be square.")
+        if colshape[0] == 0:
+            raise ValueError("Array `colcov` has invalid shape.")
```
Above 3 error code paths not covered by tests, if it matters.

johnmdusel replied:
Added coverage in b1916d6.

tylerjereddy commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,631 @@ def entropy(self):
+            iw_samples[idx], lower=True, check_finite=False
+        ).reshape(iw_samples.shape[1:])
+        # except ValueError:
+        #     import pdb; pdb.set_trace()
```
minor: leftover debug comments

johnmdusel replied:
Resolved in e45cb56.

tylerjereddy reviewed (commented):
Looks like you made a solid effort to follow our usual practices. I added a few surface-level comments ahead of the review by `stats` regulars/domain experts.

Commit:
MAINT: stats: Make checks for dimensions equal to zero more consistent and remove leftover debugging lines in `_multivariate.py`

Commit:
TST: stats: Add test coverage for remaining error code paths in `matrix_t_gen._process_parameters`

johnmdusel commented:
Is this PR still active? Some tests show as failed, but I’m unsure how to resolve them on my end. I’m happy to make any necessary changes to fix them.

lucascolley committed:
Merge branch 'main' into matrix-variate-t-distribution

lucascolley commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,626 @@ def entropy(self):
+
+    The probability density function for `matrix_t` is
+
+    .. math::
+        \mathcal{T}_{m,n} = \\frac{\Gamma_n \left( \\frac{\mathrm{df} + m + n - 1}{2} \\right)}{
+        \Gamma_n \left( \\frac{\mathrm{df} + n - 1}{2} \\right)\pi^{mn / 2}\left|\Sigma\\right|^{n/2}
+        \left|\Omega\\right|^{m/2}}\left|I_n + ( X - \mathrm{M} )^T \Sigma^{-1} ( X - \mathrm{M} ) \Omega^{-1}\\right|^{
+        -\\frac{\mathrm{df} + m + n - 1}{2}}
+
+    or, alternatively,
+
+    .. math::
+        \mathcal{T}_{m,n} = \\frac{\Gamma_m \left( \\frac{\mathrm{df} + m + n - 1}{2} \\right)}{
```
at a guess, this may fix most of the CI failures.

```suggestion
    .. math::

        \mathcal{T}_{m,n} = \\frac{\Gamma_n \left( \\frac{\mathrm{df} + m + n - 1}{2} \\right)}{
        \Gamma_n \left( \\frac{\mathrm{df} + n - 1}{2} \\right)\pi^{mn / 2}\left|\Sigma\\right|^{n/2}
        \left|\Omega\\right|^{m/2}}\left|I_n + ( X - \mathrm{M} )^T \Sigma^{-1} ( X - \mathrm{M} ) \Omega^{-1}\\right|^{
        -\\frac{\mathrm{df} + m + n - 1}{2}}

    or, alternatively,

    .. math::

        \mathcal{T}_{m,n} = \\frac{\Gamma_m \left( \\frac{\mathrm{df} + m + n - 1}{2} \\right)}{
```

johnmdusel replied:
Thanks for the tip. Docstrings are updated and and docs build/render correctly locally.

lucascolley reviewed (commented):
the lint failures shown at https://github.com/scipy/scipy/actions/runs/14981775469/job/42087379243?pr=22925 are an easy fix—they are just due to some lines being too long

Commit:
DOC: stats: Reorganize docstrings for CI lint test

lucascolley commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,657 @@ def entropy(self):
+
+    The probability density function for `matrix_t` is
+
+    .. math::
+        \mathcal{T}_{m,n} = \\frac{
```
```suggestion
    .. math::

        \mathcal{T}_{m,n} = \\frac{
```

lucascolley commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,657 @@ def entropy(self):
+
+    or, alternatively,
+
+    .. math::
+        \mathcal{T}_{m,n} = 
```
```suggestion
    .. math::

        \mathcal{T}_{m,n} = 
```

lucascolley committed:
Update _multivariate.py

lucascolley committed:
Update _multivariate.py

dschmitz89 reviewed (changes requested):
This looks very promising, thank you @johnmdusel . Could you add references to the paper in the docstring?

About tests: how can we evaluate that this is indeed the matrix t distribution? Can we test the PDF method against some external reference? Or is there a special case where this reduces to a distribution we already have implemented?

johnmdusel commented:
> This looks very promising, thank you @johnmdusel . Could you add references to the paper in the docstring?

Yes, I will add the reference to Gupta and Nagar.

> About tests: how can we evaluate that this is indeed the matrix t distribution? Can we test the PDF method against some external reference? Or is there a special case where this reduces to a distribution we already have implemented?

When I was developing this class, I tested the `rvs` and `pdf` and `logpdf` methods against Mathematica's `MatrixTDistribution`. The `matrix_t_gen` implemented here does agree with the Mathematica version. I can add some or all of that validation to the `TestMatrixT` class, if there's interest.

Commit:
DOC: stats: Add reference to `marix_t_gen` docstring

lucascolley commented:
it looks like some commits were accidentally included with a rebase

johnmdusel commented:
> it looks like some commits were accidentally included with a rebase

Apologies for that, and thanks for fixing it.

lucascolley commented:
> thanks for fixing it.

I don't think that it is fixed: the PR is still showing as 26 commits

dschmitz89 commented:
> > This looks very promising, thank you @johnmdusel . Could you add references to the paper in the docstring?
> 
> Yes, I will add the reference to Gupta and Nagar.
> 

Thanks!

> > About tests: how can we evaluate that this is indeed the matrix t distribution? Can we test the PDF method against some external reference? Or is there a special case where this reduces to a distribution we already have implemented?
> 
> When I was developing this class, I tested the `rvs` and `pdf` and `logpdf` methods against Mathematica's `MatrixTDistribution`. The `matrix_t_gen` implemented here does agree with the Mathematica version. I can add some or all of that validation to the `TestMatrixT` class, if there's interest.

That would be great. I do onot have Mathematica but I think at least @mdhaber had access in case we want to counter check the reference values.

johnmdusel commented:
> I don't think that it is fixed: the PR is still showing as 26 commits

Having difficulty but still working on removing the unwanted commits from 13 May. I will hold off on updating tests per @dschmitz89's request until they're gone.

lucascolley committed:
Merge branch 'main' into matrix-variate-t-distribution

lucascolley commented:
> until they're gone

gone now :) you just want to make sure you have pulled the latest changes before committing anything else

Commit:
MAINT: stat: Update escape character in docstring to fix failing CI tests

johnmdusel commented:
I noticed many `DeprecationWarning: invalid escape sequence '\m'`(also `\G`) while the docs were building.

mdhaber commented:
> I do not have Mathematica...

I can execute code if you can provide it

johnmdusel commented:
> I can execute code if you can provide it

For the record, here's Mathematica info for my system:

- Version 13.0.0 for Linux x86 (64-bit) (December 3, 2021)

- Release ID 13.0.0.0 (7522564, 2021120311723)

- Patch Level 0

I generated matrix t-distribution samples and their PDF values using these commands

```
mu={{1,2,3},{4,5,6}};
sigma={{1,0.5},{0.5,1}};
omega={{1,0.3,0.2},{0.3,1,0.4},{0.2,0.4,1}};
df=5;
sampleSize=1000000;
SeedRandom[42];
dist=MatrixTDistribution[mu,sigma,omega,alpha];
samples=RandomVariate[dist,sampleSize];
Export["matrixt_samples.csv",Flatten[#]&/@samples,"CSV"];
pdfs=PDF[dist,#]&/@samples;
Export["matrixt_pdfs.csv",pdfs,"CSV"];
```

I saved the first 10,000 values from each CSV into a `.npz` file with fields `"samples"` and `"pdfs`. My unit tests use that `.npz` file. Let me know if I should use the one you generated instead.

Commit:
TST: stat: Add tests against Mathematica matrix t

rgommers commented:
> I saved the first 10,000 values from each CSV into a `.npz` file with fields `"samples"` and `"pdfs`. My unit tests use that `.npz` file. Let me know if I should use the one you generated instead.

We can't commit such a large `.npz` file to the repo. Perhaps use it for testing and then remove it, leaving only the code comment about how to reproduce the testing for someone with access to Mathematica.

dschmitz89 commented:
@johnmdusel Sorry if you interpreted my request in a way that we would need that many test cases. I thought of just a handful of reference values computed by Mathematica. What we typically do is put the code for reference value creation  into a comment and then add a few accuracy tests using pytest parametrization.

Commit:
TST/MAINT: stats: Correct loading of reference values from Mathematica.

- Remove `'matrixt_samples_pdfs_from_mathematica.npz` from `stats/tests/data` and from `stats/tests/data/meson.build`

- Remove `Path` import from `stats/tests/test_multivariate.py`

- Add Mathematica code to produce reference values in `stats.tests.test_multivariate::TestMatrixT`

- Add literal values for matrix t samples and their PDF values to `stats.tests.test_multivariate::TestMatrixT.test_pdf_against_mathematica`

- Remove test `stats.tests.test_multivariate::TestMatrixT.test_rvs_against_mathematica`

dschmitz89 committed:
Merge branch 'main' into matrix-variate-t-distribution

Commit:
:MAINT: Update docstrings

Docstrings containing LaTeX are now raw.

johnmdusel commented:
Running `dev.py doc` produces errors that seem unrelated to this PR. See line 55 in attached file.
[doc-build.log](https://github.com/user-attachments/files/20411805/doc-build.log)

melissawm commented:
@johnmdusel we recently updated the build infrastructure of scipy to use `spin build` instead of dev.py - it looks like main was merged into your branch so you may need to use this command instead of `dev.py build`.

Commit:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

Commit:
MAINT: Update docstrings

Docstrings containing LaTeX are now raw.

johnmdusel commented:
The [failing check above](https://github.com/scipy/scipy/actions/runs/15397216680/job/43325858387?pr=22925) looks to be about the `match=""` argument on [line 242 of test_lsq_linear.py](https://github.com/scipy/scipy/blame/f2d22b2308bc0269e18fe2f27fe6ab17540a4ac9/scipy/optimize/tests/test_lsq_linear.py#L242). 

I can fix this but want to confirm that the intent is to verify that an `AssertionError` is raised, regardless of the message. Can anyone confirm? Thank you!

Commit:
MAINT: optimize: Remove `match=""` argument from `test_sparse_ill_conditioned`

Caught by CI check "Linux tests / Python-debug & ATLAS & sdist+wheel, fast, py3.12/npMin, pip+pytest"

Assuming the intent of this line is to verify that an `AssertionError` is raised, regardless of the message. Rather than an `AssertionError` being raised with empty content.

dschmitz89 commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,685 @@ def entropy(self):
+        if mean is not None:
+            if meanshape[0] != numrows:
+                raise ValueError(
+                    "Arrays `mean` and `rowcov` must have the " "same number of rows."
```
Copy paste error?

dschmitz89 commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,685 @@ def entropy(self):
+
+        dims = (numrows, numcols)
+
+        if df is None:
```
Note to self: this is the same default as for the `multivariate_t`.

johnmdusel replied:
Acknowledged

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,315 @@ def test_samples(self):
+        V = 0.7 * np.identity(num_cols) + np.full((num_cols, num_cols), 0.3)
+
+        # Nonpositive degrees of freedom
+        assert_raises(ValueError, matrix_t, 0, 1, 1, 0)
+        assert_raises(ValueError, matrix_t, 0, 1, 1, -1)
```
These and all following `assert_raises` need to check the error message using the `match=` clause from pytest.

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,315 @@ def test_samples(self):
+        assert_equal(matrix_t(mean=M, rowcov=Us, colcov=Vs, df=df).colcov,
+                     0.1*Ic)
+
+    def test_frozen_matrix_t(self):
+        for i in range(1,5):
+            for j in range(1,5):
```
Can this be done using
`@pytest.mark.parametrize("i", (range(1, 5))`
`@pytest.mark.parametrize("j", (range(1, 5))`

instead of the two loops?

dschmitz89 replied:
Also just three test cases should suffice instead of 4 to save some testing time.

johnmdusel replied:
Done

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,315 @@ def test_samples(self):
+    def relative_error(vec1: np.ndarray, vec2: np.ndarray):
+        numerator = np.linalg.norm(vec1 - vec2) ** 2
+        denominator = np.linalg.norm(vec1) ** 2 + np.linalg.norm(vec2) ** 2
+        return 2 * numerator / denominator
```
Why the factor of 2?

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,315 @@ def test_samples(self):
+        return 2 * numerator / denominator
+
+    @staticmethod
+    def matrix_divergence(mat_true: np.ndarray,
+                            mat_est: np.ndarray) -> float:
+        det_true = np.linalg.det(mat_true)
+        det_est = np.linalg.det(mat_est)
+        if (det_true <= 0) or (det_est <= 0):
+            return np.inf
+        trace_term = np.trace(np.linalg.inv(mat_est) @ mat_true)
+        log_detratio = np.log(det_est / det_true)
+        return (trace_term + log_detratio - len(mat_true)) / 2
+
```
Is this necessary to compute a norm in matrix space? Seems very involved if it is only needed for testing. Also explicity inverses should be avoided if possible.

johnmdusel replied:
I streamlined the calculation to use `_PSD`, which also avoids explicitly taking an inverse. This calculation is useful for comparing random variates because it is invariant under scale and the row-spread and column-spread matrices (see below) are [unique to to a positive real constant](https://reference.wolfram.com/language/ref/MatrixTDistribution.html).

dschmitz89 reviewed (changes requested):
Sorry for taking a bit longer @johnmdusel. Overall, the implementation looks good. A few comments inline and one more request regarding testing: can we add a test that for a 1D matrix it produces the same output (pdf, entropy) as the `multivariate_t` ? We typically do that to check if two implementations are the same when they should be mathematically.

johnmdusel committed:
MAINT/TST/DOC: stats: Update unit tests to address reviewer comments.

- TST: Adds test for 1D matrix t distribution agreeing with multivariate t
distribution.

- TST: Streamlines implementation of `matrix_divergence` used to compare random variates. Also avoids explicitly calculating inverses.

- MAINT: Removes `entropy` method from `matrix_t_gen`.

- DOC: Adds note about `df==1` corresponding to the matrix variate Cauchy.

johnmdusel committed:
MAINT: stats: Terminology change for `matrix_t` parameters. Formatting.

Rename `rowcov` and `colcov` to `row_spread` and `col_spread`.
Reformat code with `black`.

johnmdusel commented on scipy/stats/_multivariate.py:
````diff
@@ -1448,6 +1449,685 @@ def entropy(self):
+colcov : array_like, optional
+    Among-column covariance matrix of the distribution (default: ``1```)
+df : scalar, optional
+    Degrees of freedom (default: ```1```)
````
Done

johnmdusel commented:
@dschmitz89 Thank you for requesting the comparison with `multivariate_t`. 
While writing this test I realized that there's a mathematical subtlety in the way the matrix t versus the multivariate t are parameterized. Because of this I'm proposing a change to the parameter names for `matrix_t`. 

I'll explain the issue first:

The matrix t ($p \times m$) is parameterized by two matrices usually denoted $\Sigma (p \times p)$ and $\Omega (m \times m)$. These are raw 2nd order central moments. On the other hand, the multivariate t is parameterized by a covariance matrix--which is the normalized raw 2nd order central moment--usually denoted $U$ and called `shape` in the `multivariate_t` implementation. The matrix normal is also parameterized by covariance matrices $U (p \times p)$ and $V(m \times m)$, called `rowcov` and `colcov` in the `matrix_normal` implementation.

This difference could result in confusion, so I'm going to change the terminology used in `matrix_t` from `rowcov` and `colcov` to `row_spread` and `col_spread`. Gupta and Nagar (2000) on p.134 attribute these terms to a paper "Subjective-probability assessment methods for multivariate-t and matrix-t models" by JM Dickey, AP Dawid, JB Kadane (1986). I haven't found a digital copy, yet. 

It's now a feature that the interface for `matrix_t` doesn't match the ones for `multivariate_t` and `matrix_normal`.

johnmdusel committed:
MAINT: stats: Fix Lint / Lint (3.11) job

johnmdusel commented:
@lucascolley -- Sorry but I've made a mess of my history again. It all started when I tried to put another change into [c5df254](https://github.com/scipy/scipy/pull/22925/commits/c5df25484b3bc6207c5b9a15cbc35d1cda29e0af). Can you work your magic trick again?

dschmitz89 commented:
> @dschmitz89 Thank you for requesting the comparison with `multivariate_t`. While writing this test I realized that there's a mathematical subtlety in the way the matrix t versus the multivariate t are parameterized. Because of this I'm proposing a change to the parameter names for `matrix_t`.
> 
> I'll explain the issue first:
> 
> The matrix t ( p × m ) is parameterized by two matrices usually denoted Σ ( p × p ) and Ω ( m × m ) . These are raw 2nd order central moments. On the other hand, the multivariate t is parameterized by a covariance matrix--which is the normalized raw 2nd order central moment--usually denoted U and called `shape` in the `multivariate_t` implementation. The matrix normal is also parameterized by covariance matrices U ( p × p ) and V ( m × m ) , called `rowcov` and `colcov` in the `matrix_normal` implementation.
> 
> This difference could result in confusion, so I'm going to change the terminology used in `matrix_t` from `rowcov` and `colcov` to `row_spread` and `col_spread`. Gupta and Nagar (2000) on p.134 attribute these terms to a paper "Subjective-probability assessment methods for multivariate-t and matrix-t models" by JM Dickey, AP Dawid, JB Kadane (1986). I haven't found a digital copy, yet.
> 
> It's now a feature that the interface for `matrix_t` doesn't match the ones for `multivariate_t` and `matrix_normal`.

Puh, I have never used these distributions myself, so I do not have a strong opinion. Are you saying that the old naming `rowcov` would be wrong for `matrix_t` while it is correct for `matrix_normal`?

johnmdusel commented:
Like you said, `rowcov` and `colcov` are correct for `matrix_normal`. It's not wrong to use them for `matrix_t`, either.  [Mathematica's documentation](https://reference.wolfram.com/language/ref/MatrixTDistribution.html) calls them "row/column covariance matrix". 

IMO calling them "spread matrices" is more descriptive, especially in the context of `matrix_normal` and `multivariate_t`.

lucascolley committed:
Merge branch 'main' into HEAD

[skip ci]

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,427 @@ def test_samples(self):
+
+        atol = 1e-10
+
+        samples_m = np.array(
+            [
+                [
+                    [0.6399716994253741, 2.171718671534955, 2.5758260933527706],
+                    [4.031082477912233, 5.0216809585266375, 6.268126154787008],
+                ],
+                [
+                    [1.1648428842062322, 2.5262970999930445, 3.7813752298650685],
```
Does Mathematica have a function to convert its arbitrary precision numbers to common double precision? These numbers have up to 17 digits but double precision typically only yields up to 15 significant digits.

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1429,6 +1429,427 @@ def test_samples(self):
+        assert_allclose(m, mT.T, atol=atol)
+        assert_allclose(m.T, mT, atol=atol)
+
+    def test_against_multivariate_t(self):
+        r"""
+        Gupta and Nagar (2000) p.133f
+        When the number of rows or the number of columns equals 1 the
+        matrix t reduces to the multivariate t. But, the matrix t
+        is parameterized by raw 2nd moments whereas the multivariate t is
+        parameterized by a covariance (raw 2nd central moment normalized by df).
+        We can see the difference by comparing the author's notation
+            $t_p(n, \omega, \mathbf{\mu}, \Sigma)$
+        for a matrix t with a single column
+        to the formula (4.1.2) for the PDF of the multivariate t.
+        """
```
Thanks for the explanation :). Could you parametrize the test so that it also covers both cases (1 column and 1 row)?

johnmdusel replied:
Always glad to write about math :). Done (af59a7ae5).

dschmitz89 committed:
Merge branch 'main' into matrix-variate-t-distribution

johnmdusel committed:
MAINT: stats: Adjust numbers from Mathematica to use double precision.

johnmdusel committed:
TST: stats: Parametrize TestMatrixT.test_against_multivariate_t to cover 1 column and 1 row cases

johnmdusel committed:
MAINT: stats: Remove unnecessary commented out lines

Commit:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

dschmitz89 commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,609 @@ def entropy(self):
+    When :math:`\mathrm{df} = 1` this distribution is known as the matrix
+    variate Cauchy.
+
+    .. versionadded:: 1.16.0
```
```suggestion
    .. versionadded:: 1.17.0
```

dschmitz89 commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,609 @@ def entropy(self):
+        std_norm = random_state.standard_normal(
+            size=(dims[1], size, dims[0])
+        ).transpose(1, 2, 0)
+        if dims[0] <= dims[1]:
+            rowchol = _cholesky_invwishart_rvs(df, row_spread, size, random_state)
+            colchol = scipy.linalg.cholesky(col_spread, lower=True)[np.newaxis, ...]
+        else:
+            rowchol = scipy.linalg.cholesky(row_spread, lower=True)[np.newaxis, ...]
+            colchol = _cholesky_invwishart_rvs(df, col_spread, size, random_state)
+        t_raw = np.einsum("ijp,ipq,ikq->ijk", rowchol, std_norm, colchol, optimize=True)
```
Could you add some explanatory comments for this code or a reference?

dschmitz89 reviewed (changes requested):
A few more comments mostly about the interface @johnmdusel. Would you have time to also have a look @mdhaber (especially the mathematica reference)?

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
DOC: stats: Update docstrings per reviewer feedback.

- Example style: inspections in their own lines

- Add note about multivariate gamma function

- Update `versionadded`

johnmdusel committed:
BUG: stats: Remove logic that silently changes nans to an identity matrix

johnmdusel committed:
DOC: stats: Add explanation of equivalent PDF expressions

Motivated by their role in the `rvs` method

johnmdusel commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,609 @@ def entropy(self):
+    >>> import numpy as np
+    >>> from scipy.stats import matrix_t
+    >>> M = np.arange(6).reshape(3,2); M
+    array([[0, 1],
```
Done (7dc12e61ff8ffab7654daa83f1c18d2aeb729170) 

Originally, I was following the style o fthe [`matrix_normal` example](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.matrix_normal.html#scipy.stats.matrix_normal)

```
>>> M = np.arange(6).reshape(3,2); M
array([[0, 1],
       [2, 3],
       [4, 5]])
>>> U = np.diag([1,2,3]); U
array([[1, 0, 0],
       [0, 2, 0],
       [0, 0, 3]])
>>> V = 0.3*np.identity(2); V
array([[ 0.3,  0. ],
       [ 0. ,  0.3]])
>>> X = M + 0.1; X
array([[ 0.1,  1.1],
       [ 2.1,  3.1],
       [ 4.1,  5.1]])
>>> matrix_normal.pdf(X, mean=M, rowcov=U, colcov=V)
0.023410202050005054
```

johnmdusel commented on scipy/stats/_multivariate.py:
```diff
@@ -1448,6 +1449,609 @@ def entropy(self):
+    where :math:`\mathrm{M}` is the mean,
+    :math:`\Sigma` is the row-wise spread matrix,
+    :math:`\Omega` is the column-wise matrix,
+    and :math:`\mathrm{df}` is the degrees of freedom.
```
Done (7dc12e61ff8ffab7654daa83f1c18d2aeb729170)

johnmdusel committed:
DOC: stats: Fix docstrings for CircleCI tests

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
MAINT: stats: Change default values in `matrix_t_gen.__call__` for interface consistency

Post `0dceecb` they need to be `row_spread=1` and `col_spread=1` instead of `None`. This is how they're defined in `matrix_normal_gen.__call__`

mdhaber commented:
> Would you have time to also have a look @mdhaber (especially the mathematica reference)?

I'm limited to using Mathematica version "14.2.1 for Linux x86 (64-bit) (March 16, 2025)". When I execute that code, I don't get the sample sample or pdf values.

dschmitz89 commented:
> > Would you have time to also have a look @mdhaber (especially the mathematica reference)?
> 
> I'm limited to using Mathematica version "14.2.1 for Linux x86 (64-bit) (March 16, 2025)". When I execute that code, I don't get the sample sample or pdf values.

Hm, unfortunate. Can we do a minimal test with a fixed set of matrices to test the pdf values if we cannot rely on the random variates? I do not expect differences in the first few digits between different mathematica versions. @johnmdusel  @mdhaber

mdhaber commented:
>

johnmdusel committed:
TST: stats: Add test of PDF values against Julia `MatrixTDist`

johnmdusel commented:
@dschmitz89 @mdhaber I added a unit test that checks PDF values against Julia. The source in the test's docstring and this [small repo](https://github.com/johnmdusel/matrix-t-julia). All tests pass locally.

dschmitz89 reviewed (approved):
I am happy with this PR and the pdf values were also confirmed against an external source (Mathematica). There is one last stylistic nit regarding tests: we typically use relative tolerances, not absolute ones. Will wait for other reviewers before merging.

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
MAINT/TST: stats: Change absolute tolerances to relative tolerances where appropriate

johnmdusel commented:
@rgommers The tests against [Julia](https://github.com/johnmdusel/scipy/blob/d861da77ec67b3bd2d46fc31f1955a5a3b9bc393/scipy/stats/tests/test_multivariate.py#L1703) and [Mathematica](https://github.com/johnmdusel/scipy/blob/d861da77ec67b3bd2d46fc31f1955a5a3b9bc393/scipy/stats/tests/test_multivariate.py#L1800) are now based on a handful of reference values, instead of a file.

dschmitz89 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1892,7 +1892,7 @@ def test_samples(self):
         V = 0.7 * np.identity(num_cols) + np.full((num_cols, num_cols), 0.3)
         N = 10**4
-        atol = 1e-1
+        rtol = 1e-1
```
A relative tolerance of 10% is quite high, can we lower it?

johnmdusel replied:
I set the relative tolerance to 1%, trading off on the sample size.

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
MAINT/TST: stats: Lower relative tolerance for transpose check

johnmdusel committed:
Merge branch 'main' into matrix-variate-t-distribution

johnmdusel commented:
@dschmitz89 -- questions about the failing CI checks above

- [Linux tests](https://github.com/scipy/scipy/actions/runs/16263981032/job/46069887969?pr=22925): The `test_samples` test took too long; it takes more samples for the test to pass with a lower relative tolerance. How would you like me to manage this tradeoff?

- [Windows tests](https://github.com/scipy/scipy/actions/runs/16263981027/job/46069887230?pr=22925): This seems to be related to HiGHS; can I assume it'll be fixed in some other PR?

dschmitz89 committed:
Merge branch 'main' into matrix-variate-t-distribution

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

johnmdusel committed:
MAINT/TST: stats: Set relative tolerance for transpose check to 5%

johnmdusel commented:
> > @dschmitz89 -- questions about the failing CI checks above
> > 
> > * [Linux tests](https://github.com/scipy/scipy/actions/runs/16263981032/job/46069887969?pr=22925): The `test_samples` test took too long; it takes more samples for the test to pass with a lower relative tolerance. How would you like me to manage this tradeoff? With 10k samples it will pass with a 5% relative tolerance; with 100k it will pass with 2.5% but will take > 1 second.
> 
> Let's go with the 5% tolerance for simplicity. Sry about the churn.

No problem! I made the change.

johnmdusel commented:
> @rgommers The tests against [Julia](https://github.com/johnmdusel/scipy/blob/d861da77ec67b3bd2d46fc31f1955a5a3b9bc393/scipy/stats/tests/test_multivariate.py#L1703) and [Mathematica](https://github.com/johnmdusel/scipy/blob/d861da77ec67b3bd2d46fc31f1955a5a3b9bc393/scipy/stats/tests/test_multivariate.py#L1800) are now based on a handful of reference values, instead of a file.

@rgommers would you be able to approve if no further changes are required? On the other hand, I'm happy to make further  changes as needed. Thank you very much!

johnmdusel committed:
Merge remote-tracking branch 'upstream/main' into matrix-variate-t-distribution

lucascolley committed:
Merge branch 'main' into matrix-variate-t-distribution

dschmitz89 commented:
Thanks for your first (and large) contribution @johnmdusel !
